### ISSUE TITLE
Improve LockService functionality and flexibility

### ISSUE THREAD

**Issue by nvoxland** [liquibase/liquibase#1453]
## Description

Because concurrent update/rollback/etc. operations to a database will interfere with each other, Liquibase has a [LockService](https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/lockservice/LockService.java) which we use to coordinate access so only one instance can run against a database at time.

Examples of when you can run into concurrency trouble:
- Liquibase is embedded on startup of an application server, and you have a cluster of machines that may be starting at once
- Liquibase runs as part of a build job, and concurrent builds can happen at once

The [StandardLockService](https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/lockservice/StandardLockService.java) is the normal implementation that uses a `databasechangeloglock` table. Liquibase inserts a record and commits it at startup time, and if other instances see that record they wait until it is cleared.

The big problem with this standard implementation is that IF the instance doing the update is killed before it commits the clearing of the record, that value is stuck until `liquibase clearLocks` is ran. 

This problem has been around forever in Liquibase, but has been exasperated in recent years with systems like AWS that auto-kill processes that seem slow to start up. We wrap everything in try/catch blocks, but process killing happens too out-of-band for us to deal with.

## Related issues/PRs

Descriptions and proposed fixes:
- #829 
- #1417 
- #1311 
- Others?

#### BYO Locking

One option is to leave the locking to the users. There are many situations where Liquibase is ran within a system where the user knows only one instance can be ran at a time. For example, cloud providers have an "init" phase where the platform ensures just one container runs and Liquibase can be moved to there. Or, updates are done via a managed deployment script or build process which is ensuring only one version of an app is being deployed at once. 

In these cases, the lock service is just getting in the way. Ideally, the default lock service should work just fine and whether it runs unnecessarily or not isn't worth the hassle of turning it off. But, it's a fallback option to consider. We do have [the nodatabaselock](https://github.com/liquibase/liquibase-nochangeloglock) extension which could be folded into the main liquibase code with an easy flag to switch to that implementation

#### Database Session Based Locking

Databases have ways to lock tables for the duration of a connection that automatically unlocks them when the connection closes.

By using that functionality, we can rely on the database to clean up the lock. We don't even necessarily need the databasechangeloglock table anymore, because we can lock the databasechangelog table itself. 

However, this approach is going to be specific to each database since the syntax is different for each database. The semantics of what a "lock" means and the visibility of it also varies by database, so it will require extensive testing to ensure it works as expected everywhere. 

A native lock system may break the "lock has been held by X since Y" tracking we can do with the databasechangeloglock table, since that native lock ownership may not be visible to other connections. But, we could still preserve the databasechangelog table for reporting purposes even if we don't use it for locking purposes anymore.

Examples of lock syntax:
- Postgresql: pg_try_advisory_lock https://www.postgresql.org/docs/9.1/functions-admin.html
- Mysql: lock/unlock tables https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html
- Oracle: lock/unlock tables: https://docs.oracle.com/javadb/10.8.3.0/ref/rrefsqlj40506.html
- MSSQL: select with tablock https://docs.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-table?view=sql-server-ver15

#### Heartbeat

Rather than relying on a native database approach, we could have a thread within the LockService update the databasechangeloglock table every 30 seconds. If a separate process sees that the heartbeat is over 30 seconds old, it can take the lock.

This avoids the database specific logic, but also requires threads which some application servers do not like creating. But, maybe that is OK still?

#### Max Lock Time

There have been some suggestions to have a setting where the lock attempt succeeds if the old lock is older than X seconds. I tend to not like this because sometimes updates (especially the first run or when creating indexes on a large table) can just take a long time. To avoid this, the "take lock" has to be long. But, a long enough take-lock setting doesn't help the use case of cloud providers auto-killing processes and wanting to restart a new one right away. For example, a 30 minute setting may sometimes not be long enough and the lock can be incorrectly taken while at the same time way too long for people to wait for their auto-restart logic to take the lock.

#### Other options?

Any other ideas on how we could implement the logic?

## Proposed Changes

Currently, I'm thinking:
1. Bring a NoLockService implementation into liquibase-core with a way to enable it. 
2. Update StandardLockService to use a heartbeat thread to override locks. The heartbeat thread/logic can be disabled with a configuration flag for cases where it doesn't work
3. After releasing the heartbeat/nolock options, listen for issues people continue to have. If there are none, great. If there are more then look to implement the database-specific use native locking.

But, I'm looking for feedback on that idea.

**Comment by stoffel2107** [comment]
Well, the Max lock time is not a solid enough IMO.
Even though unlikely it'll always be possible to create changelogs that validly exceed that time limit and would therefore break. (Even if the time is on 1+ days theoretically it's still possible however unlikely)

BYO Locking => Well it's error-prone since the user has to manually configure this and it's very specific. Maybe a nice to have?

Database Session Lock =>This sounds reasonable but would require some work due to the mentioned database implementations

Heartbeat =>Is the best idea here IMO!

**Comment by nvoxland** [comment]
I agree that the heartbeat setup will help without getting into database-specific functionality.  I'll update the ticket to be less of a proposal and more of a definition. I have a branch started with this work and will make a PR for it soon.

**Comment by nvoxland** [comment]
I created #1541 for the in-progress work on implementing this

**Comment by nvoxland** [comment]
Another thing to keep in mind with the implementation: make sure the base classes handle multi-threading support correctly.

**Comment by nmck257** [comment]
Hey - so, given this:

> Realized an issue with the heartbeat strategy: Liquibase currently makes just one connection but the heartbeat thread needs a separate connection for transaction reasons. We can't have the heartbeat committing the normal liquibase transaction, and we can't have it waiting for a commit from the normal connection.
> 
> We'll have to build in support for opening separate connections before we can make progress on this.
@nvoxland in #1541 

Do we realistically expect to add in support for separate connections? Eg would that capability be useful for other features?

I know that Max Lock Time wasn't preferred, but given the blocker above, it certainly seems simpler to implement and deliver (and can be extended with heartbeats in the future). And as long as the max age is configurable, users can decide for themselves whether they want to accommodate long-running changelogs with a big value or rapid kill-and-reboot cloud environments with a small value.

As a user, I'd definitely take a pretty-good solution over no solution.

**Comment by RichardBradley** [comment]
The plugin https://github.com/blagerweij/liquibase-sessionlock (released mid 2020) now implements your option "Database Session Based Locking", and does it very well. It works for MySQL, Postgresql and Oracle. Personally, I think it is a much better solution than a timeout based approach, and the price of writing database specific adapters for it is worth paying. (In a way, isn't one of the primary goals of Liquibase to write database specific adapters for common operations?). In my opinion it would be nice to bring that into core as it is strictly better than the standard lock.

If you are looking at locking issues, I think double checked locking is also worth pursuing. See my comments on #829 and #2105

Thanks for Liquibase, I think it's a great project.

**Comment by nvoxland** [comment]
Thanks for the heads-up on liquibase-sessionlock. I'll take a look at that.

I'll leave this ticket open for a while until we actually bring that into core or something similar.

**Comment by stdmitry** [comment]
Can you please explain what's the purpose of locking in case of no new migrations?

In my case, one app gets restarted over and over because of infrastructure issues and eventually grabs the database lock. So even when infrastructure issues are fixed, the app is unable to start by itself. Even when there was no need for any migrations.

**Comment by kataggart** [comment]
@stdmitry check out https://github.com/liquibase/liquibase/pull/2190; people are actively working on a resolution that should address your use case (and sorry your application has to get restarted over and over again... sounds a bit painful)

**Comment by RichardBradley** [comment]
> Can you please explain what's the purpose of locking in case of no new migrations?

The purpose of locking when examining the changelogs table is so that the changelog state cannot be invalidated by another copy of the app running in parallel that may make changes.

There is a PR at #2190 that will skip this lock if there are no new migrations to run.

It's worth noting that liquibase will still need to take a lock if there are migrations to run, even after #2190, so if your app is frequently killed unexpectedly, you will still see orphaned locks. The best fix for that is a locking mechanism which will auto-unlock when the client is killed, like https://github.com/blagerweij/liquibase-sessionlock . As per my comment above, I think liquibase should pull that plugin into core, as it is better than the current lock implementation with no downsides and addresses the orphaned lock issue, which is a very common complaint with Liquibase. I am not aware of an issue or PR to track that work other than this.

**Comment by michelecurioni** [comment]
> > Can you please explain what's the purpose of locking in case of no new migrations?
> 
> The purpose of locking when examining the changelogs table is so that the changelog state cannot be invalidated by another copy of the app running in parallel that may make changes.
> 
> There is a PR at #2190 that will skip this lock if there are no new migrations to run.
> 
> It's worth noting that liquibase will still need to take a lock if there are migrations to run, even after #2190, so if your app is frequently killed unexpectedly, you will still see orphaned locks. The best fix for that is a locking mechanism which will auto-unlock when the client is killed, like https://github.com/blagerweij/liquibase-sessionlock . As per my comment above, I think liquibase should pull that plugin into core, as it is better than the current lock implementation with no downsides and addresses the orphaned lock issue, which is a very common complaint with Liquibase. I am not aware of an issue or PR to track that work other than this.



> Thanks for the heads-up on liquibase-sessionlock. I'll take a look at that.
> 
> I'll leave this ticket open for a while until we actually bring that into core or something similar.

I just wanted to call out that some connection pooling mechanisms use a transaction based pooling, which would not work with the session based locks.

**Comment by kailasandhale** [comment]
Are we still considering resolving this? It seems it's open for last 2 years.

**Comment by bsofiato** [comment]
Are you looking for someone to implement this ? If no one is working on this I can try to create a PR for this :)
